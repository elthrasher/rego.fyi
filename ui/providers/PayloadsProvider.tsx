import React, { createContext, ReactNode, useContext, useState } from 'react';

const baseUrl = 'https://api.rego.fyi';

const dataDefault = JSON.stringify({
  requests: [{ methods: ['GET'], resources: ['/orders'] }],
  permissions: ['start_order', 'view_invoice'],
  subscriptions: ['newsletter'],
});

const regoDefault = `package entitlements

import data.requests
import data.permissions
import data.subscriptions

default allow = false

allow {
    check_entitlements[input]
}

check_entitlements[input] {
    r = requests[_]
    some i; match_with_wildcard(data.permissions, input.permissions[i])
    some j; match_with_wildcard(data.subscriptions, input.subscriptions[j])
    match_with_wildcard(r.methods, input.method)
    match_with_wildcard(r.resources, input.resource)
}

match_with_wildcard(allowed, value) {
    allowed[_] = "*"
}
match_with_wildcard(allowed, value) {
    allowed[_] = value
}`;

const tokenDefault = JSON.stringify({
  permissions: ['start_order'],
  resource: '/orders',
  method: 'GET',
  subscriptions: ['newsletter'],
});

export interface Payloads {
  data: string;
  rego: string;
  token: string;

  method: string;
  resource: string;

  status?: 'ALLOWED' | 'DENIED';
  responsePayload?: string;

  saveData: (data: string) => string;
  saveRego: (data: string) => string;
  saveToken: (data: string) => string;
  sendRequest: () => void;
  updateRequest: (method: string, resource: string) => void;
}

export const PayloadsContext = createContext<Payloads | undefined>(undefined);

export const PayloadsProvider = ({ children }: { children: ReactNode }): JSX.Element => {
  const [data, setData] = useState<string>(dataDefault);
  const [rego, setRego] = useState<string>(regoDefault);
  const [token, setToken] = useState<string>(tokenDefault);

  const [method, setMethod] = useState<string>('GET');
  const [resource, setResource] = useState<string>('/orders');

  const [status, setStatus] = useState<'ALLOWED' | 'DENIED' | undefined>();
  const [responsePayload, setResponsePayload] = useState<string | undefined>();

  const saveData = (data: string): string => {
    setData(data);
    try {
      JSON.parse(data);
      return 'valid';
    } catch (e) {
      return 'Invalid JSON';
    }
  };

  const saveRego = (rego: string): string => {
    setRego(rego);
    return 'valid';
  };

  const saveToken = (data: string): string => {
    setToken(data);
    try {
      JSON.parse(data);
      return 'valid';
    } catch (e) {
      return 'Invalid JSON';
    }
  };

  const updateRequest = (method: string, resource: string) => {
    setMethod(method);
    setResource(resource);
  };

  const sendRequest = async () => {
    const pl = btoa(JSON.stringify({ data, rego, token }));
    try {
      const response = await fetch(`${baseUrl}${resource}`, { headers: { Authorization: `Bearer ${pl}` }, method });
      if (response.status === 200) {
        setStatus('ALLOWED');
        const body = await response.json();
        setResponsePayload(JSON.stringify(body));
      } else {
        setStatus('DENIED');
        setResponsePayload(undefined);
      }
    } catch {
      setStatus('DENIED');
      setResponsePayload(undefined);
    }
  };

  return (
    <PayloadsContext.Provider
      value={{
        data,
        rego,
        token,
        method,
        resource,
        saveData,
        saveRego,
        saveToken,
        sendRequest,
        updateRequest,
        status,
        responsePayload,
      }}
    >
      {children}
    </PayloadsContext.Provider>
  );
};

export const usePayloadsContext = (): Payloads => {
  const payloadsContext = useContext(PayloadsContext);
  if (!payloadsContext) {
    throw new Error('usePayloadsContext must be used within the PayloadsContext.Provider');
  }
  return payloadsContext;
};
