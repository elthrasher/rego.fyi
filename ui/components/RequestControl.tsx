import Container from '@material-ui/core/Container';
import FormControl from '@material-ui/core/FormControl';
import Grid from '@material-ui/core/Grid';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import Select from '@material-ui/core/Select';
import TextField from '@material-ui/core/TextField';
import React, { ChangeEvent } from 'react';
import { usePayloadsContext } from '../providers/PayloadsProvider';

export const RequestControl = (): JSX.Element => {
  const { method, resource, updateRequest } = usePayloadsContext();

  const onSelectChange = (
    event: ChangeEvent<{
      name?: string | undefined;
      value: unknown;
    }>,
  ) => updateRequest(event.target.value as string, resource);

  const onTextChange = (
    event: ChangeEvent<{
      name?: string | undefined;
      value: unknown;
    }>,
  ) => updateRequest(method, event.target.value as string);

  return (
    <Container>
      <Grid container spacing={3}>
        <Grid item xs={6}>
          <FormControl>
            <InputLabel>Method</InputLabel>
            <Select onChange={onSelectChange} value={method}>
              <MenuItem value={'DELETE'}>DELETE</MenuItem>
              <MenuItem value={'GET'}>GET</MenuItem>
              <MenuItem value={'PATCH'}>PATCH</MenuItem>
              <MenuItem value={'POST'}>POST</MenuItem>
              <MenuItem value={'PUT'}>PUT</MenuItem>
            </Select>
          </FormControl>
        </Grid>
        <Grid item xs={6}>
          <FormControl>
            <TextField label="Resource" onChange={onTextChange} value={resource} />
          </FormControl>
        </Grid>
      </Grid>
    </Container>
  );
};
