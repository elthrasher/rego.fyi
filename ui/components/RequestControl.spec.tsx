import React from 'react';
import renderer from 'react-test-renderer';

import { PayloadsProvider } from '../providers/PayloadsProvider';
import { RequestControl } from './RequestControl';

it('renders a message', () => {
  const component = renderer.create(
    <PayloadsProvider>
      <RequestControl />
    </PayloadsProvider>,
    {
      createNodeMock: (node) => {
        return document.createElement(node.type.toString());
      },
    },
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  //  const instance = component.getInstance();
  //  instance.
});
