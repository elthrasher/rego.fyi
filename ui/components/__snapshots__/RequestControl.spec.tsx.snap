// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a message 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthLg"
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <div
        className="MuiFormControl-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
          data-shrink={true}
        >
          Method
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <div
            aria-haspopup="listbox"
            className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
            onBlur={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            role="button"
            tabIndex={0}
          >
            GET
          </div>
          <input
            aria-hidden={true}
            className="MuiSelect-nativeInput"
            onAnimationStart={[Function]}
            onChange={[Function]}
            required={false}
            tabIndex={-1}
            value="GET"
          />
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSelect-icon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <div
        className="MuiFormControl-root"
      >
        <div
          className="MuiFormControl-root MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink={false}
          >
            Resource
          </label>
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiInput-input"
              disabled={false}
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
              value="/orders"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
